Grant/contracts/BuilderGrant.sol

```diff
@@ -0,0 +1,350 @@
+        season.grants[0].amount = grantsAmount[0];
+        season.grants[1].amount = grantsAmount[1];
+        season.grants[2].amount = grantsAmount[2];
+        season.claimStartedAt = uint40(block.timestamp);
+
+        // Reset ranker data if exists, and overwrites the old data
+        if (season.rankers.length > 0) {
+            delete season.rankers;
+        }
+        unchecked {
+            for (uint256 i = 0; i < fids.length; ++i) {
+                season.rankers.push(Ranker(fids[i], wallets[i], 0, [false, false, false]));
+            }
+        }
+
+        emit SetSeasonData(seasonId, fids.length, grantsAmount);
+    }
+
+    function getClaimableAmountByTop3(
+        uint256 seasonId,
+        uint256 ranking
+    ) public view validSeasonId(seasonId) returns (uint16) {
+        // Only the top 3 rankers (index: 0, 1, 2) can claim
+        if (ranking > 2) revert InvalidRankingParam();
+
+        // Check claim deadline
+        Season storage season = seasons[seasonId];
+        if (season.claimStartedAt + TOP3_CLAIM_DEADLINE < block.timestamp) revert ClaimDeadlineReached();
+
+        // Check if the grant has not been claimed yet
+        if (season.grants[ranking].claimedType != 0) revert AlreadyClaimed();
+
+        return season.grants[ranking].amount;
+    }
+
+    function claimByTop3(uint256 seasonId, uint256 ranking, uint8 claimType) external validSeasonId(seasonId) {
+        // Basic validations are done here
+        uint16 claimableAmount = getClaimableAmountByTop3(seasonId, ranking);
+
+        Season storage season = seasons[seasonId];
+        address msgSender = _msgSender();
+
+        // Check msg.sender is the winner
+        if (season.rankers[ranking].wallet != msgSender) revert PermissionDenied();
+
+        // claimType - 0: not claimed yet, 1: 100% self, 2: 50% donation, 3: 100% donation
+        uint16 amountForSelf;
+        uint16 amountForDonation;
+        if (claimType == 1) {
+            // 100% self
+            amountForSelf = claimableAmount;
+        } else if (claimType == 2) {
+            // 50% donation
+            amountForSelf = amountForDonation = claimableAmount / 2;
+        } else if (claimType == 3) {
+            // 100% donation
+            amountForDonation = claimableAmount;
+        } else {
+            revert InvalidClaimType();
+        }
+
+        // Set ranker claimed
+        season.grants[ranking].claimedType = claimType;
+
+        // Send Mini Buildings reward
+        if (amountForSelf > 0) {
+            season.rankers[ranking].claimedAmount += amountForSelf;
+            season.totalClaimed += amountForSelf;
+            if (!_mintBuildings(amountForSelf, msgSender)) revert MintBuildingsFailed();
+        }
+
+        // Set donations
+        if (amountForDonation > 0) {
+            unchecked {
+                for (uint256 i = 3; i < amountForDonation + 3; ++i) {
+                    season.rankers[i].donationReceived[ranking] = true;
```

Comment:
GAS: You could reduce gas and storage requirements by just tracking `season.amountForDonation[3]` and then simplifying line 276 to a simple `ranking <= season.amountForDuration[ranking]` check

Reply:
Oh, this is because we want to keep track of where the donations came from. For example, if the 10th ranker got 2 donations from the 1st and 3rd rankers, we want to display them later so the donation receiver can show appreciation to the givers.

Reply:
Oh I see... If I'm not mistaken, you could still figure this out with the approach I propose. 

It would be a little more work because you'd be tracking the *number* of donation recipients per donor 0/1/2, so the data would be there, but you'd have check each Ranker's ranking vs this number rather than a simple lookup 👍

Reply:
Oh, I hadn’t thought about that method. Grants and rankers data are not meant to be updated, so you are right! 

It might make the lookup a bit more complicated, and our front-end devs are almost finished with their work, so I may just keep the current structure for now :)

Thanks for pointing this out!

Oh, and please leave your public wallet address for receiving the bounty!